### ISSUE TITLE
create_graph failed for custom CUDA extension when calling torch.autograd.grad

### ISSUE THREAD

**Issue by 731894915** [pytorch/pytorch#39922]
When using torch.autograd.grad to calculate gradient, we hope pyotrch can create graph for gradient with create_graph sets to _True_. This works fine for custom subclass of torch.autograd.Function which you can verify here [https://gist.github.com/Daiver/1d702b5a9faf0a22e3007500c243da93](url).

However, torch.autograd.grad() would fail to create graph (gradient calculation is correct) if there are CUDA extensions. The CUDA extension is called PreciseRoIPooling [https://github.com/vacancy/PreciseRoIPooling](url). Here is a tiny example.

```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchviz import make_dot
from PreciseRoIPooling.pytorch.prroi_pool import PrRoIPool2D

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.pool = PrRoIPool2D(3, 3, spatial_scale=0.5)
        self.f1 = nn.Linear(9,1)

    def forward(self, features,rois):
        # Run the IoUNet module
        poolfeat = self.pool(features, rois)
        poolfeat = poolfeat.reshape(2, 16, -1)
        iou = self.f1(poolfeat)
        out = iou * iou
        g_out = make_dot(out)

        # calculate gradient w.r.t nodes behind CUDA extension
        grad_poolfeat= torch.autograd.grad(out, poolfeat, grad_outputs=torch.ones_like(out), retain_graph=True, create_graph=True)[0]
        g_grad_poolfeat = make_dot(grad_poolfeat)
       # calculate gradient w.r.t nodes before CUDA extension
        grad_roi = torch.autograd.grad(grad_poolfeat, rois, grad_outputs=torch.ones_like(grad_poolfeat), retain_graph=True, create_graph=True)
        g_grad_roi = make_dot(grad_roi)

        return iou

if __name__ == '__main__':
    model = Model().to('cuda:0')
    features = torch.rand((4, 16, 32, 32),requires_grad=True).cuda()
    rois = torch.tensor([
        [0, 0, 0, 14, 14],
        [1, 14, 14, 28, 28],
    ]).float().cuda()
    rois.requires_grad=True
    iou = model(features, rois)
```

### Computation graphs
I use torchviz to draw the computation graph.

1. Fisrt, the computation graph of out:
![image](https://user-images.githubusercontent.com/17667639/84460441-82597d80-ac9c-11ea-9652-d6bea0ca1bcf.png)

2. And after I calculate gradient w.r.t some node that behind CUDA extension(backward does not go through this extension), the computation graph looks like this. Everything is fine. PyTorch creates graph for output gradient.
![image](https://user-images.githubusercontent.com/17667639/84460620-feec5c00-ac9c-11ea-80c5-dd3bc6f30532.png)

3. However, when I trying to calculate gradient w.r.t nodes before CUDA extension(backward does go through this extension), create graph fails. Output of torch.autograd.grad has no grad_fn and does not require gradient. Computation graph becomes _None_.
![image](https://user-images.githubusercontent.com/17667639/84460901-b08b8d00-ac9d-11ea-8e9d-587b0d8eb0b5.png)

**Update**
Here are the codes and outputs of custom autograd function.
```
@staticmethod
    def backward(ctx, grad_output):
        pydevd.settrace(suspend=False, trace_only_current_thread=True)
        print('---grad_output status---')
        print('requires_grad:', grad_output.requires_grad)
        print('grad_fn:', grad_output.grad_fn)

        _prroi_pooling = _import_prroi_pooling()

        features, rois, output = ctx.saved_tensors
        grad_input = grad_coor = None

        if features.requires_grad:
            grad_output = grad_output.contiguous()
            grad_input = _prroi_pooling.prroi_pooling_backward_cuda(features, rois, output, grad_output, *ctx.params)
            print('---grad_input status---')
            print('requires_grad:', grad_input.requires_grad)
            print('grad_fn:', grad_input.grad_fn)
        if rois.requires_grad:
            grad_output = grad_output.contiguous()
            grad_coor = _prroi_pooling.prroi_pooling_coor_backward_cuda(features, rois, output, grad_output, *ctx.params)
            print('---grad_coor status---')
            print('requires_grad:', grad_input.requires_grad)
            print('grad_fn:', grad_input.grad_fn)

        return grad_input, grad_coor, None, None, None
```
I printed out _requires_grad_ and _grad_fn_ of `grad_output` and `grad_input`. As shown in the results, `grad_output` works fine. But after we call the CUDA extension，calculated gradients(`grad_input` and `grad_coor`) lose their _grad_fn_. 

```
---grad_output status---
requires_grad: True
grad_fn: <ViewBackward object at 0x7fc7891d6e10>
---grad_input status---
requires_grad: False
grad_fn: None
---grad_coor status---
requires_grad: False
grad_fn: None
```

I also tried to manipulate `grad_output` within python environment such as make a new tensor _out_ and `out = grad_output * grad_output` instead of using CUDA extension. This new tensor also gets correct _grad_fn_(<MulBackward0 object at 0x7f4bf042dc90>)

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Corresponding forum post: https://discuss.pytorch.org/t/torch-autograd-grad-returns-gradient-with-none-grad-fn-even-set-create-graph-true/85049
